### DISCUSSION TITLE
[5일차 - 윤경] 깊은복사 vs 얕은 복사

### DISCUSSION THREAD

**Discussion by skyhanull** [zzangCS/zzangInterview#20] [프론트 면접 질문]
깊은 복사와 얕은 복사의 차이는?

**Comment by skyhanull** [comment]
깊은 복사 vs 얕은 복사

**얕은 복사(Shallow Copy)**란 객체를 복사할 때 **원본 값과 복사된 값이 같은 참조(=메모리 주소)를 가리키는 것**이다.

**깊은 복사(Deep Copy)**
는 **복사된 객체가 다른 주소를 참조**하며 **내부의 값만 복사**된다.

**Comment by ez1n** [comment]
- **원시값 vs 참조값**

⇒ 원시값 : 값을 복사할 때 복사된 값을 다른 메모리에 할당하기 때문에 원래의 값과 복사된 값이 서로에게 영향을 미치지 않는다.
⇒ 참조값 : 변수가 객체의 주소를 가리키는 값이기 때문에 복사된 값(주소)이 같은 값을 가리킨다.

- **얕은 복사 (Shallow Copy)**

> 객체를 복사했을 때 기존 값과 복사된 값이 같은 참조(주소)를 가리키고 있는 것

⇒ Array.prototype.slice() : 2차원 배열인 경우 얕은 복사를 수행한다.
⇒ Object.assign()
⇒ Spread 연산자
    
- **깊은 복사 (Deep Copy)**

> 객체 안에 객체가 있을 경우에도 원본과의 참조가 완전히 끊어진 객체

⇒ JSON.stringify() && JSON.parse() : 객체를 json 문자열로 변환하는 과정에서 원본 객체와의 참조가 모두 끊어진다. 이후 JSON.parse()를 이용하여 다시 자바스크립트 객체로 만든다.
⇒ 재귀 함수 이용

**Reply by ez1n** [reply]
https://bbangson.tistory.com/78

**Comment by smarfy99** [comment]
얕은 복사 : 참조 타입(reference) 데이터가 저장한 메모리 주소 값을 복사한 것
깊은 복사 : 새로운 메모리 공간을 확보해 완전히 복사한 것

얕은 복사 시, 해당 변수를 재사용하여 수정하면 원본 값이 동시에 변한다.
